<a href="https://colab.research.google.com/github/MinervaNunez/Proyecto-de-Python-BEDU/blob/Rama-Jesus/Proyecto_Equipo_8_Kevin_Nicky_Jesus.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Proyecto de Python Equipo 8** 

# Adquisición de datos


### Obtención de los datos mediante API

El gobierno de México utiliza el sistema de repositorios de datos abiertos [CKAN](https://datos.gob.mx/blog/ckan?category=api-cdn&tag=educacion) para compartir los Datos Abiertos de las dependencias de la Administración Pública Federal, así como de gobiernos locales.

Utilizando la API de CKAN es posible consultar los detalles de los conjuntos de datos que se ofrecen en el portal datos.gob.mx/busca

De esta forma, mediante peticiones a esta API podemos obtener la URL del conjunto de datos.

In [43]:
import pandas as pd
import requests
import numpy as np

El endpoint para consultar datos es:

https://datos.gob.mx/busca/api/3/action/package_search?q=BUSQUEDA 

Por lo que pasaremos el valor "desaparecidos" al parametro q, para obtener los datasets relacionados.

In [44]:
endpoint = 'https://datos.gob.mx/busca/api/3/action/package_search?'
payload = {'q': 'desaparecidos'}

In [45]:
r = requests.get(endpoint, params=payload)

In [46]:
r.status_code

200

| La petición se ha realizado con éxito

In [47]:
json = r.json()

In [48]:
json.keys()

dict_keys(['help', 'success', 'result'])

La llave *'result'* es la que nos interesa, exploraremos su contenido para identificar el conjunto de datos que buscamos.

In [49]:
for key in json['result']['results']:
    print(key['name'])

estadisticas-de-apoyo-y-proteccion-a-detenidos-y-desaparecidos
registro-nacional-de-datos-de-personas-extraviadas-o-desaparecidas-rnped
registro-nacional-de-datos-de-personas-extraviadas-o-desaparecidas-rnped-sitio-espejo-3
registro-nacional-de-datos-de-personas-extraviadas-o-desaparecidas-rnped-sitio-espejo-2
registro-nacional-de-datos-de-personas-extraviadas-o-desaparecidas-rnped-sitio-espejo-1
registro-nacional-de-datos-de-personas-extraviadas-o-desaparecidas-rnped-sitio-espejo-4
cantidad-de-derechohabientes-beneficiados-por-el-programa-integral-de-becas-para-hijos-de-milit


| El conjunto de datos que buscamos se encuentra en *'registro-nacional-de-datos-de-personas-extraviadas-o-desaparecidas-rnped'* con indice 1.

Ahora bsucaremos el conjunto de datos dentro de este diccionario.

In [50]:
for key in json['result']['results'][1]['resources']:
    print(f"Nombre del Dataset: {key['name']} \nURL: {key['url']} \n")

Nombre del Dataset: Base de datos del RNPED del fuero común 
URL: https://secretariadoejecutivo.gob.mx/transparencia/datos_abiertos/RNPEDFC.csv 

Nombre del Dataset: Base de datos del RNPED del fuero federal 
URL: https://secretariadoejecutivo.gob.mx/docs/datos_abiertos/RNPEDFF__1.csv 



| Nuestro conjunto de datos es *'Base de datos del RNPED del fuero común'* con indice 0.

**Guardamos la URL de nuestro conjunto y leemos el archivo .csv y Guardamos la versión RAW de los datos en fromato csv y leemos el archivo**

In [51]:
url = json['result']['results'][1]['resources'][0]['url']

**En caso de error de conexion por falla del servidor leemos el dataset de nuestra direccion de GITHUB**

In [52]:
try:
    raw = pd.read_csv(url, encoding='ISO-8859-1')
    raw.to_csv('RNPEDFC.csv', index=False, encoding='ISO-8859-1')
    df = pd.read_csv('RNPEDFC.csv', encoding='ISO-8859-1')
except:
  import pandas as pd
  df= pd.read_csv('https://raw.githubusercontent.com/MinervaNunez/Proyecto-de-Python-BEDU/Rama-Jesus/RNPEDFC.csv', sep=',', encoding='ISO-8859-1')

**Obtenermos como resultado nuesto data frame principal**

In [53]:
df

Fecha en que se le vio por ultima vez  \
0                                2012-03-18   
1                                2007-02-05   
2                                2009-03-14   
3                                2006-12-23   
4                                2008-01-11   
...                                     ...   
36260                            2018-04-26   
36261                            2018-01-08   
36262                            2015-02-02   
36263                            2017-08-22   
36264                            2018-04-11   

      Hora en que se le vio por ultima vez  \
0                                 05:00:00   
1                                 02:00:26   
2                                 08:30:54   
3                                 08:00:28   
4                                 12:00:17   
...                                    ...   
36260                             08:00:00   
36261                             11:00:00   
36262                             07:00:00   
36263                             04:00:00   
36264                             02:00:00   

      Pais en que se le vio por ultima vez  \
0                                   MEXICO   
1                                   MEXICO   
2                                   MEXICO   
3                                   MEXICO   
4                                   MEXICO   
...                                    ...   
36260                               MEXICO   
36261                               MEXICO   
36262                               MEXICO   
36263                               MEXICO   
36264                               MEXICO   

      Entidad en que se le vio por ultima vez  \
0                                      SONORA   
1                            ESTADO DE MEXICO   
2                            ESTADO DE MEXICO   
3                            ESTADO DE MEXICO   
4                            ESTADO DE MEXICO   
...                                       ...   
36260                                  PUEBLA   
36261                                  COLIMA   
36262                                  PUEBLA   
36263                                  COLIMA   
36264                              NUEVO LEON   

      Municipio en que se le vio por ultima vez  \
0                                       NAVOJOA   
1                                VALLE DE BRAVO   
2                                        CHALCO   
3                           TLALNEPANTLA DE BAZ   
4                                        TOLUCA   
...                                         ...   
36260                                    PUEBLA   
36261                          VILLA DE ALVAREZ   
36262                                     CHILA   
36263                          VILLA DE ALVAREZ   
36264                  SAN NICOLAS DE LOS GARZA   

      Localidad en que se le vio por ultima vez Nacionalidad Estatura  \
0                                       NAVOJOA     MEXICANA      1.1   
1                                VALLE DE BRAVO     MEXICANA      1.1   
2                                        CHALCO     MEXICANA      1.1   
3                           TLALNEPANTLA DE BAZ     MEXICANA      1.1   
4                                        TOLUCA     MEXICANA      1.1   
...                                         ...          ...      ...   
36260                                    PUEBLA     MEXICANA      1.6   
36261                          VILLA DE ALVAREZ     MEXICANA      1.7   
36262                      CHILAS DE LAS FLORES     MEXICANA     1.77   
36263                          VILLA DE ALVAREZ     MEXICANA     1.76   
36264                  SAN NICOLAS DE LOS GARZA     MEXICANA      1.6   

      Complexion    Sexo Edad  \
0        DELGADA   MUJER    7   
1        DELGADA   MUJER    4   
2        DELGADA  HOMBRE    6   
3        DELGADA  HOMBRE   10   
4        DELGADA  HOMBRE    4   
...          ...     ...  ...   
36260    DELGADA  HOMBRE 

# Análisis exploratorio de datos

Seguimos la metodología vista en clase.

Primero, obtenemos las dimensiones de nuestro `DataFrame`.

In [54]:
df.shape

(36265, 15)

La celda anterior nos dice que nuestro dataframe tiene 36265 renglones (observaciones), y 15 columnas (variables).

Veamos cuáles son nuestras variables.

In [55]:
df.columns

Index(['Fecha en que se le vio por ultima vez',
       'Hora en que se le vio por ultima vez',
       'Pais en que se le vio por ultima vez',
       'Entidad en que se le vio por ultima vez',
       'Municipio en que se le vio por ultima vez',
       'Localidad en que se le vio por ultima vez', 'Nacionalidad', 'Estatura',
       'Complexion', 'Sexo', 'Edad', 'Descripcion de senas particulares',
       'Etnia', 'Discapacidad', 'Dependencia que envio la informacion'],
      dtype='object')

In [56]:
df.dtypes

Fecha en que se le vio por ultima vez        object
Hora en que se le vio por ultima vez         object
Pais en que se le vio por ultima vez         object
Entidad en que se le vio por ultima vez      object
Municipio en que se le vio por ultima vez    object
Localidad en que se le vio por ultima vez    object
Nacionalidad                                 object
Estatura                                     object
Complexion                                   object
Sexo                                         object
Edad                                         object
Descripcion de senas particulares            object
Etnia                                        object
Discapacidad                                 object
Dependencia que envio la informacion         object
dtype: object

De acuerdo con el resultado de la anterior celda, nuestras 15 variables y sus tipos de datos (por defecto) son:
    
| Variable | Tipo |
| -------  | ---- |
| Fecha en que se le vio por ultima vez        | object |
| Hora en que se le vio por ultima vez         | object |
| Pais en que se le vio por ultima vez         | object |
| Entidad en que se le vio por ultima vez      | object |
| Municipio en que se le vio por ultima vez    | object |
| Localidad en que se le vio por ultima vez    | object |
| Nacionalidad                                 | object |
| Estatura                                     | object |
| Complexion                                   | object |
| Sexo                                         | object |
| Edad                                         | object |
| Descripcion de senas particulares            | object |
| Etnia                                        | object |
| Discapacidad                                 | object |
| Dependencia que envio la informacion         | object |


In [57]:
df.index

RangeIndex(start=0, stop=36265, step=1)

Del anterior comando, `pandas` nos dice que nuestras observaciones están indizadas por valores numéricos que empiezan desde el 0 y terminan en 36265.

Finalmente, echamos un vistazo a los datos para darnos una mejor idea de cómo está compuesto nuestro `DataFrame` a analizar.

In [58]:
df.head(5)

Fecha en que se le vio por ultima vez Hora en que se le vio por ultima vez  \
0                            2012-03-18                             05:00:00   
1                            2007-02-05                             02:00:26   
2                            2009-03-14                             08:30:54   
3                            2006-12-23                             08:00:28   
4                            2008-01-11                             12:00:17   

  Pais en que se le vio por ultima vez  \
0                               MEXICO   
1                               MEXICO   
2                               MEXICO   
3                               MEXICO   
4                               MEXICO   

  Entidad en que se le vio por ultima vez  \
0                                  SONORA   
1                        ESTADO DE MEXICO   
2                        ESTADO DE MEXICO   
3                        ESTADO DE MEXICO   
4                        ESTADO DE MEXICO   

  Municipio en que se le vio por ultima vez  \
0                                   NAVOJOA   
1                            VALLE DE BRAVO   
2                                    CHALCO   
3                       TLALNEPANTLA DE BAZ   
4                                    TOLUCA   

  Localidad en que se le vio por ultima vez Nacionalidad Estatura Complexion  \
0                                   NAVOJOA     MEXICANA      1.1    DELGADA   
1                            VALLE DE BRAVO     MEXICANA      1.1    DELGADA   
2                                    CHALCO     MEXICANA      1.1    DELGADA   
3                       TLALNEPANTLA DE BAZ     MEXICANA      1.1    DELGADA   
4                                    TOLUCA     MEXICANA      1.1    DELGADA   

     Sexo Edad                  Descripcion de senas particulares  \
0   MUJER    7  TIENE UNA CICATRIZ EN LA RODILLA IZQUIERDA POR...   
1   MUJER    4                                    NO ESPECIFICADO   
2  HOMBRE    6  MANCHAS, PARTE SUPERIOR DE PIERNA DERECHA, CIC...   
3  HOMBRE   10  LUNARES, A LA ALTURA DE LA CINTURAFORMA DE MAN...   
4  HOMBRE    4                                    NO ESPECIFICADO   

             Etnia Discapacidad Dependencia que envio la informacion  
0  NO ESPECIFICADO      NINGUNO                         FGE - SONORA  
1  NO ESPECIFICADO      NINGUNO               FGJ - ESTADO DE MEXICO  
2  NO ESPECIFICADO      NINGUNO               FGJ - ESTADO DE MEXICO  
3  NO ESPECIFICADO      NINGUNO               FGJ - ESTADO DE MEXICO  
4  NO ESPECIFICADO      NINGUNO               FGJ - ESTADO DE MEXICO

In [59]:
df.tail(10)

Fecha en que se le vio por ultima vez  \
36255                            2018-04-25   
36256                            2018-04-22   
36257                            2017-07-02   
36258                            2017-04-06   
36259                            2018-04-15   
36260                            2018-04-26   
36261                            2018-01-08   
36262                            2015-02-02   
36263                            2017-08-22   
36264                            2018-04-11   

      Hora en que se le vio por ultima vez  \
36255                             06:30:00   
36256                             11:30:00   
36257                             11:14:59   
36258                             02:00:00   
36259                             06:30:00   
36260                             08:00:00   
36261                             11:00:00   
36262                             07:00:00   
36263                             04:00:00   
36264                             02:00:00   

      Pais en que se le vio por ultima vez  \
36255                               MEXICO   
36256                               MEXICO   
36257                               MEXICO   
36258                               MEXICO   
36259                               MEXICO   
36260                               MEXICO   
36261                               MEXICO   
36262                               MEXICO   
36263                               MEXICO   
36264                               MEXICO   

      Entidad en que se le vio por ultima vez  \
36255                                  PUEBLA   
36256                                  PUEBLA   
36257                                  SONORA   
36258                                  COLIMA   
36259                                  COLIMA   
36260                                  PUEBLA   
36261                                  COLIMA   
36262                                  PUEBLA   
36263                                  COLIMA   
36264                              NUEVO LEON   

      Municipio en que se le vio por ultima vez  \
36255                     SAN MARTIN TEXMELUCAN   
36256                                    PUEBLA   
36257                                   CANANEA   
36258                                MANZANILLO   
36259                                    COLIMA   
36260                                    PUEBLA   
36261                          VILLA DE ALVAREZ   
36262                                     CHILA   
36263                          VILLA DE ALVAREZ   
36264                  SAN NICOLAS DE LOS GARZA   

      Localidad en que se le vio por ultima vez Nacionalidad         Estatura  \
36255                     SAN MARTIN TEXMELUCAN     MEXICANA              1.7   
36256                                    PUEBLA     MEXICANA             1.52   
36257                                   CANANEA     MEXICANA  NO ESPECIFICADO   
36258                                  SANTIAGO     MEXICANA             1.71   
36259                                    COLIMA     MEXICANA             1.59   
36260                                    PUEBLA     MEXICANA              1.6   
36261                          VILLA DE ALVAREZ     MEXICANA              1.7   
36262                      CHILAS DE LAS FLORES     MEXICANA             1.77   
36263                          VILLA DE ALVAREZ     MEXICANA             1.76   
36264                  SAN NICOLAS DE LOS GARZA     MEXICANA              1.6   

            Complexion    Sexo             Edad  \
36255          DELGADA  HOMBRE               26   
36256          DELGADA   MUJER               18   
36257  NO ESPECIFICADO  HOMBRE  NO ESPECIFICADO   
36258          ROBUSTA  HOMBRE               22   
36259          ROBUSTA  HOMBRE               57   
36260          DELGADA  HOMBRE               22   
36261          ROBUSTA   MUJER               27   
36262          MEDIANA   MUJER               86   
36263          ROBUST

Del vistazo a los últimos 10 valores de nuestro dataset podemos observar que nuestro dataset contiene algunos datos faltantes etiquetados por el valor `NO ESPECIFICADO`.

¿Será que existe alguna observación para la cual el pais donde se le vió por última vez sea distinto de México?
Para contestar esta incógnita podemos ver si la cantidad de registros que satisfacen que `df['Pais en que se le vio por ultima vez'] != 'MEXICO']` es mayor a 0

In [60]:
len(df.loc[df['Pais en que se le vio por ultima vez'] != 'MEXICO']) > 0

False

Como la instrucción anterior retornó `False`, entonces el País en el que se le vio por última vez siempre es México.

¿Cuáles son todos los posibles valores que puede tomar la variable `Entidad en que se le vio por ultima vez`?

In [61]:
def val_unicos_col(df, col_name):
    return set(df[col_name])

col_name = 'Entidad en que se le vio por ultima vez'

val_col_entidad = val_unicos_col(df, col_name)
val_col_entidad

{'AGUASCALIENTES',
 'BAJA CALIFORNIA',
 'BAJA CALIFORNIA SUR',
 'CAMPECHE',
 'CHIAPAS',
 'CHIHUAHUA',
 'CIUDAD DE MEXICO',
 'COAHUILA DE ZARAGOZA',
 'COLIMA',
 'DURANGO',
 'ESTADO DE MEXICO',
 'GUANAJUATO',
 'GUERRERO',
 'HIDALGO',
 'JALISCO',
 'MICHOACAN',
 'MORELOS',
 'NAYARIT',
 'NO ESPECIFICADO',
 'NUEVO LEON',
 'OAXACA',
 'PUEBLA',
 'QUERETARO',
 'QUINTANA ROO',
 'SAN LUIS POTOSI',
 'SINALOA',
 'SONORA',
 'TABASCO',
 'TAMAULIPAS',
 'TLAXCALA',
 'VERACRUZ',
 'YUCATAN',
 'ZACATECAS'}

¿Cuáles son todos los posibles valores que puede tomar la variable `Municipio en que se le vio por ultima vez`?

In [62]:
col_name = 'Municipio en que se le vio por ultima vez'

val_col_municipio = val_unicos_col(df, col_name)
val_col_municipio
len(val_col_municipio)

1068

Vemos que son muchos los municipios, por lo que mejor preguntamos por la cantidad de municipios únicos. La respuesta a esta pregunta es 1068 municipios únicos.


¿Cuáles son todos los posibles valores que puede tomar la variable `Localidad en que se le vio por ultima vez`?

In [63]:
col_name = 'Localidad en que se le vio por ultima vez'

val_col_localidad = val_unicos_col(df, col_name)
val_col_localidad
len(val_col_localidad)

2676

De nuevo, vemos que son muchas las localidades, por lo que mejor preguntamos por la cantidad de valores únicos que toma la variable. La respuesta a esta pregunta es 2676 localidades únicas.


Vale la pena preguntarse si la variable `Nacionalidad` toma valores distintos de `MEXICANA`.

In [64]:
col_name = 'Nacionalidad'

val_col_nacionalidad = val_unicos_col(df, col_name)
len(val_col_nacionalidad)

26

Como la cantidad de elementos distintos en la columna es 26, entonces podemos afirmar que en efecto la variable `Nacionalidad` toma valores distintos de `MEXICANA`.

De igual forma podemos investigar qué valores únicos puede tomar la variable `Complexion`.

In [65]:
col_name = 'Complexion'
val_col_complexion = val_unicos_col(df, col_name)
val_col_complexion

{'DELGADA',
 'MEDIANA',
 'NO ESPECIFICADO',
 'No Especificado',
 'OBESA',
 'ROBUSTA'}

De lo anterior podemos observar que en particular en nuestro dataset hay dos formas de especificar que un valor es `NA`:
Puede ser escrito como `"NO ESPECIFICADO"` o como `No Especificado`. Habría que manejar estas dos formas de escribir el valor en la limpieza de los datos.

¿Hay más registros con valor `No Especificado`?

In [66]:
## TODO

## Checar si no hay "NO ESPECIFICADO"s escritos de forma distinta.

Podemos preguntarnos si el campo `Sexo` permite el registro de gente no binaria.

In [67]:
col_name = 'Sexo'
val_col_sexo = val_unicos_col(df, col_name)
val_col_sexo

{'HOMBRE', 'MUJER'}

Por lo tanto, la variable `Sexo` sólo admite dos valores: `HOMBRE` y `MUJER`.

Veamos cuál es el rango de edad:

In [68]:
col_name = 'Edad'
val_col_edad = val_unicos_col(df, col_name)
val_col_edad

lista = list(
    map(
        lambda x: int(x) if x.isdigit() else -1, 
        val_col_edad
    )
)
lista.sort()

print(f'El mínimo es {lista[1]} año(s) y el máximo es de {lista[len(lista)-1]} año(s)')

El mínimo es 1 año(s) y el máximo es de 103 año(s)


Entonces, el rango de edad va desde 1 año hasta 103 años.

Ahora investiguemos cuál es el rango de alturas:

In [69]:
def isfloat(s):
    try:
        float(s)
        
        return True
    except:
        return False

col_name = 'Estatura'
val_col_estatura = val_unicos_col(df, col_name)
val_col_estatura

lista = list(
    map(
        lambda x: float(x) if isfloat(x) else -1, 
        val_col_estatura
    )
)
lista.sort()
lista = list(
    filter(lambda x: x != -1, lista)
)

print(f'El mínimo de altura es de {lista[0]} centimetros y el máximo es de altura es de {lista[len(lista)-1]} centimetros')

El mínimo de altura es de 0.3 centimetros y el máximo es de altura es de 170.0 centimetros


Por lo tanto, el rango de altura es de 0.3 cm a 170.0 cm.

Cabe destacar que hubieron valores `NO ESPECIFICADO` (`NA`s), por lo que deben limpiarse según convenga.

Los valores que toma la variable `Etnia` son

In [70]:
col_name = 'Etnia'

val_col_etnia = val_unicos_col(df, col_name)
val_col_etnia

{'AMUZGOS OAXACA',
 'CHATINOS',
 'CHICHIMECAS',
 'CHINANTECOS',
 'CHOLES',
 'HUASTECOS',
 'HUICHOLES',
 'MAMES',
 'MAYAS',
 'MEXICANEROS',
 'MIXTECOS',
 'NAHUAS',
 'NO ESPECIFICADO',
 'OTOMIES',
 'PAMES',
 'SERIS',
 'TARAHUMARAS - RARAMURI',
 'TLALPANECOS',
 'TOTONACAS',
 'YAQUIS'}

Otra variable de interés es la variable `Discapacidad`.

In [71]:
col_name = 'Discapacidad'

val_col_discapacidad = val_unicos_col(df, col_name)
val_col_discapacidad

{'AUTISMO', 'NINGUNO', 'NO ESPECIFICADO', 'SINDROME DE DOWN'}

Del bloque anterior podemos apreciar que son 4 posibles valores: `AUTISMO`, `NINGUNO`, `NO ESPECIFICADO` y `SINDROME DE DOWN`

Finalmente, procedemos a investigar la variable `Dependencia que envio la informacion`

In [72]:
col_name = 'Dependencia que envio la informacion'

val_col_dependencia = val_unicos_col(df, col_name)
val_col_dependencia

{'FGE - AGUASCALIENTES',
 'FGE - CAMPECHE',
 'FGE - CHIAPAS',
 'FGE - CHIHUAHUA',
 'FGE - COAHUILA',
 'FGE - DURANGO',
 'FGE - GUERRERO',
 'FGE - JALISCO',
 'FGE - MORELOS',
 'FGE - NAYARIT',
 'FGE - OAXACA',
 'FGE - PUEBLA',
 'FGE - QUERETARO',
 'FGE - QUINTANA ROO',
 'FGE - SAN LUIS POTOSI',
 'FGE - SINALOA',
 'FGE - SONORA',
 'FGE - TABASCO',
 'FGE - VERACRUZ',
 'FGE - YUCATAN',
 'FGE - ZACATECAS',
 'FGJ - ESTADO DE MEXICO',
 'PGJ - BAJA CALIFORNIA',
 'PGJ - BAJA CALIFORNIA SUR',
 'PGJ - CIUDAD DE MEXICO',
 'PGJ - COLIMA',
 'PGJ - GUANAJUATO',
 'PGJ - HIDALGO',
 'PGJ - MICHOACAN',
 'PGJ - NUEVO LEON',
 'PGJ - TAMAULIPAS',
 'PGJ - TLAXCALA'}

### **Conclusiones del primer análisis exploratorio de datos**

Nuestro dataset está constituido de 36265 observaciones de 15 variables indizadas del 0 al 36264. Nuestras variables son

| Variable |
| -------  |
| Fecha en que se le vio por ultima vez        |
| Hora en que se le vio por ultima vez         |
| Pais en que se le vio por ultima vez         |
| Entidad en que se le vio por ultima vez      |
| Municipio en que se le vio por ultima vez    |
| Localidad en que se le vio por ultima vez    |
| Nacionalidad                                 |
| Estatura                                     |
| Complexion                                   |
| Sexo                                         |
| Edad                                         |
| Descripcion de senas particulares            |
| Etnia                                        |
| Discapacidad                                 |
| Dependencia que envio la informacion         |

De una observación preliminar de los primeros y últimos 10 datos podemos observar que el tipo de dato inferido por pandas es erróneo (el tipo de dato inferido fue `object`).

La propuesta es la siguiente:

| Variable | Tipo |
| -------  | ---- |
| Fecha en que se le vio por ultima vez        | Fecha con formato 'yyyy-mm-dd' |
| Hora en que se le vio por ultima vez         | Hora con formato 'hh:mm:ss' |
| Pais en que se le vio por ultima vez         | Categórico (`MEXICO`) |
| Entidad en que se le vio por ultima vez      | Categórico (valores de `val_col_entidad`) |
| Municipio en que se le vio por ultima vez    | Categórica (valores de `val_col_municipio`) |
| Localidad en que se le vio por ultima vez    | Categórica (valores de `val_col_localidad`) |
| Nacionalidad                                 | Categórica (valores de `val_col_nacionalidad`) |
| Estatura                                     | Numérico float64 (`0.3` a `170.0`) |
| Complexion                                   | Categórica (`DELGADA`, `MEDIANA`, `NO ESPECIFICADA`, `OBESA`, `ROBUSTA`) |
| Sexo                                         | Categórica (`HOMBRE`, `MUJER`) |
| Edad                                         | Numérico int64 (`1` a `103`) |
| Descripcion de senas particulares            | object (string) |
| Etnia                                        | Categórica (valores de  `val_col_etnia`)|
| Discapacidad                                 | Categórica (`AUTISMO`, `NINGUNO`, `NO ESPECIFICADO`, `SINDROME DE DOWN`) |
| Dependencia que envio la informacion         | Categórica (valores de `val_col_dependencia`) |

# **Transformación y limpieza de los datos**

Conocemo los tipos de datos del Dataframe original

In [73]:
df.dtypes

Fecha en que se le vio por ultima vez        object
Hora en que se le vio por ultima vez         object
Pais en que se le vio por ultima vez         object
Entidad en que se le vio por ultima vez      object
Municipio en que se le vio por ultima vez    object
Localidad en que se le vio por ultima vez    object
Nacionalidad                                 object
Estatura                                     object
Complexion                                   object
Sexo                                         object
Edad                                         object
Descripcion de senas particulares            object
Etnia                                        object
Discapacidad                                 object
Dependencia que envio la informacion         object
dtype: object

Buscamos NANS

In [74]:
#Exploramos NAN
df.isna().sum()

Fecha en que se le vio por ultima vez        0
Hora en que se le vio por ultima vez         0
Pais en que se le vio por ultima vez         0
Entidad en que se le vio por ultima vez      0
Municipio en que se le vio por ultima vez    0
Localidad en que se le vio por ultima vez    0
Nacionalidad                                 0
Estatura                                     0
Complexion                                   0
Sexo                                         0
Edad                                         0
Descripcion de senas particulares            0
Etnia                                        0
Discapacidad                                 0
Dependencia que envio la informacion         0
dtype: int64

Como vemos que el index empieza con 0, lo reindexamos para iniciar con 1 hasta el 36265 y se lo aplicamos a un Dataframe, conservando el original


In [75]:
fin=len(df.index)
#fin
lst = list(np.arange(1,fin+1))
#lst
len(lst)
df_2=df.reindex(lst, method='ffill')
df_2.head(100)

Fecha en que se le vio por ultima vez  \
1                              2007-02-05   
2                              2009-03-14   
3                              2006-12-23   
4                              2008-01-11   
5                              2012-05-27   
..                                    ...   
96                             2008-05-03   
97                             2008-05-13   
98                             2008-07-27   
99                             2008-07-29   
100                            2008-11-01   

    Hora en que se le vio por ultima vez Pais en que se le vio por ultima vez  \
1                               02:00:26                               MEXICO   
2                               08:30:54                               MEXICO   
3                               08:00:28                               MEXICO   
4                               12:00:17                               MEXICO   
5                               12:00:43                               MEXICO   
..                                   ...                                  ...   
96                              03:00:00                               MEXICO   
97                              09:00:00                               MEXICO   
98                              11:30:00                               MEXICO   
99                              02:00:00                               MEXICO   
100                             07:00:00                               MEXICO   

    Entidad en que se le vio por ultima vez  \
1                          ESTADO DE MEXICO   
2                          ESTADO DE MEXICO   
3                          ESTADO DE MEXICO   
4                          ESTADO DE MEXICO   
5                          ESTADO DE MEXICO   
..                                      ...   
96                                  HIDALGO   
97                                  HIDALGO   
98                                  HIDALGO   
99                                  HIDALGO   
100                                 HIDALGO   

    Municipio en que se le vio por ultima vez  \
1                              VALLE DE BRAVO   
2                                      CHALCO   
3                         TLALNEPANTLA DE BAZ   
4                                      TOLUCA   
5                              NICOLAS ROMERO   
..                                        ...   
96                        TULANCINGO DE BRAVO   
97                                    ACATLAN   
98                                    ACATLAN   
99                        TULANCINGO DE BRAVO   
100                       TULANCINGO DE BRAVO   

    Localidad en que se le vio por ultima vez Nacionalidad Estatura  \
1                              VALLE DE BRAVO     MEXICANA      1.1   
2                                      CHALCO     MEXICANA      1.1   
3                         TLALNEPANTLA DE BAZ     MEXICANA      1.1   
4                                      TOLUCA     MEXICANA      1.1   
5                              NICOLAS ROMERO     MEXICANA      1.1   
..                                        ...          ...      ...   
96                                 TULANCINGO     MEXICANA      1.6   
97                                    ACATLAN     MEXICANA      1.6   
98                                    ACATLAN     MEXICANA      1.6   
99                        TULANCINGO DE BRAVO     MEXICANA      1.6   
100                       TULANCINGO DE BRAVO     MEXICANA      1.7   

    Complexion    Sexo Edad  \
1      DELGADA   MUJER    4   
2      DELGADA  HOMBRE    6   
3      DELGADA  HOMBRE   10   
4      DELGADA  HOMBRE    4   
5      DELGADA  HOMBRE    9   
..         ...     ...  ...   
96     ROBUSTA  HOMBRE   12   
97     ROBUSTA  HOMBRE   37   
98     DELGADA  HOMBRE   16   
99     DELGADA   MUJER   36   
100    ROBUSTA  HOMBRE   23   

                     Descripcion de senas particulares            Etnia  \
1                                  

Observamos el nombre de las columnas, las primeras 6 se pueden estandarizar y las demas para cumplir la convención

In [76]:
# Se contruye un mapeo con el nombre de las columnas esperados
column_name_mapping = {
    'Fecha en que se le vio por ultima vez': 'fecha',
    'Hora en que se le vio por ultima vez': 'hora',
    'Pais en que se le vio por ultima vez': 'pais',
    'Entidad en que se le vio por ultima vez': 'entidad',
    'Municipio en que se le vio por ultima vez': 'municipio',
    'Localidad en que se le vio por ultima vez': 'localidad',
    'Nacionalidad': 'nacionalidad',
    'Estatura': 'estatura',
    'Complexion': 'complexion',
    'Sexo': 'sexo',
    'Edad': 'edad',
    'Descripcion de senas particulares': 'senas_particulares',
    'Etnia': 'etnia',
    'Discapacidad': 'discapacidad',
    'Dependencia que envio la informacion': 'dependecia_origen'
    
}
df_2 = df_2.rename(columns=column_name_mapping)

for column in df_2:
    df_2[column]=df_2[column].str.title()




En nuestro Dataframe nos dimos cuenta que no tenemos NANS, sin embargo tenemos valores con la leyenda "No Especificado"

In [77]:
# Explorandovalores denominados "NO ESPECIFICADOS"
print(f'{("Columna"):42} =No Especificado=  Porcentaje\n')
for col in df_2.columns:
    total = (df_2[col] == "NO ESPECIFICADO").sum()+(df_2[col] == "No Especificado").sum()
    print(f'{(col):42} | {(total):16} {(round((total*100)/df.shape[0], 2)):9}%')

Columna                                    =No Especificado=  Porcentaje

fecha                                      |              338      0.93%
hora                                       |               16      0.04%
pais                                       |                0       0.0%
entidad                                    |               29      0.08%
municipio                                  |              668      1.84%
localidad                                  |             3413      9.41%
nacionalidad                               |             2040      5.63%
estatura                                   |            11008     30.35%
complexion                                 |            11295     31.15%
sexo                                       |                0       0.0%
edad                                       |             3156       8.7%
senas_particulares                         |            18915     52.16%
etnia                                      |      

In [78]:
from textwrap import fill
# INICIAMOS CON LA LIMPIEZA DE DATOS SEGUN LOS PORCENTAJES MOSTRADOS ARRIBA, ELIMINANDO LOS VALORES QUE NO NOS DAN VALOR

# Quitamos los no especificados del campo fecha, colocando en su lugar NAN y despues se elimina
df_2["fecha"]=df_2["fecha"].replace('No Especificado',np.NaN)
df_2=df_2.dropna(subset=["fecha"])

#Revisamos de nuevo 
print(f'{("Columna"):42} =No Especificado= \n')
for col in df_2.columns:
    print(f'{(col):42} | {(df_2[col] == "No Especificado").sum()}')


df_2.head(1000)

Columna                                    =No Especificado= 

fecha                                      | 0
hora                                       | 0
pais                                       | 0
entidad                                    | 26
municipio                                  | 635
localidad                                  | 3345
nacionalidad                               | 2028
estatura                                   | 10784
complexion                                 | 11157
sexo                                       | 0
edad                                       | 3088
senas_particulares                         | 18705
etnia                                      | 35795
discapacidad                               | 1043
dependecia_origen                          | 0


fecha      hora    pais           entidad            municipio  \
1     2007-02-05  02:00:26  Mexico  Estado De Mexico       Valle De Bravo   
2     2009-03-14  08:30:54  Mexico  Estado De Mexico               Chalco   
3     2006-12-23  08:00:28  Mexico  Estado De Mexico  Tlalnepantla De Baz   
4     2008-01-11  12:00:17  Mexico  Estado De Mexico               Toluca   
5     2012-05-27  12:00:43  Mexico  Estado De Mexico       Nicolas Romero   
...          ...       ...     ...               ...                  ...   
1020  2007-07-18  12:00:00  Mexico           Sinaloa             Culiacan   
1021  2007-04-07  12:00:00  Mexico           Sinaloa             Culiacan   
1022  2007-02-25  08:00:00  Mexico           Sinaloa             Culiacan   
1023  2007-06-02  06:00:00  Mexico           Sinaloa             Culiacan   
1024  2007-03-12  12:00:00  Mexico           Sinaloa             Culiacan   

                localidad nacionalidad         estatura       complexion  \
1          Valle De Bravo     Mexicana              1.1          Delgada   
2                  Chalco     Mexicana              1.1          Delgada   
3     Tlalnepantla De Baz     Mexicana              1.1          Delgada   
4                  Toluca     Mexicana              1.1          Delgada   
5          Nicolas Romero     Mexicana              1.1          Delgada   
...                   ...          ...              ...              ...   
1020             Culiacan     Mexicana  No Especificado  No Especificado   
1021             Culiacan     Mexicana  No Especificado  No Especificado   
1022           Costa Rica     Mexicana  No Especificado  No Especificado   
1023             Culiacan     Mexicana  No Especificado  No Especificado   
1024             Culiacan     Mexicana  No Especificado  No Especificado   

        sexo             edad  \
1      Mujer                4   
2     Hombre                6   
3     Hombre               10   
4     Hombre                4   
5     Hombre                9   
...      ...              ...   
1020  Hombre               41   
1021  Hombre  No Especificado   
1022  Hombre               23   
1023  Hombre               38   
1024  Hombre               60   

                                     senas_particulares            etnia  \
1                                       No Especificado  No Especificado   
2     Manchas, Parte Superior De Pierna Derecha, Cic...  No Especificado   
3     Lunares, A La Altura De La Cinturaforma De Man...  No Especificado   
4                                       No Especificado  No Especificado   
5                                       No Especificado  No Especificado   
...                                                 ...              ...   
1020                                    No Especificado  No Especificado   
1021                                    No Especificado  No Especificado   
1022                                    No Especificado  No Especificado   
1023                                    No Especificado  No Especificado   
1024                                    No Especificado  No Especificado   

     discapacidad       dependecia_origen  
1         Ninguno  Fgj - Estado De Mexico  
2         Ninguno  Fgj - Estado De Mexico  
3         Ninguno  Fgj - Estado De Mexico  
4         Ninguno  Fgj - Estado De Mexico  
5         Ninguno  Fgj - Estado De Mexico  
...           ...                     ...  
1020      Ninguno           Fge - Sinaloa  
1021      Ninguno           Fge - Sinaloa  
1022      Ninguno           Fge - Sinaloa  
1023      Ninguno           Fge - Sinaloa  
1024      Ninguno           Fge - Sinaloa  

[1000 rows x 15 columns]

In [101]:
# UNA VEZ ELIMINADOS LOS VALORES NAN PROCEDEMOS A PASAR LAS COLUMNAS A LOS TIPOS DE DATOS CORRECTOS

# Convertivos el campo de fecha
df_2['fecha']= pd.to_datetime(df_2['fecha'],errors='coerce')

#Convertimos la serie de edad a numerico con pd.to_numeric, inidicando que si encuentra error que continue, pero los pone como NAN
df_2['edad']= pd.to_numeric(df_2['edad'],errors='coerce')
#rellenamos los NAN de Edad con la media
mean_value=df_2['edad'].mean() 
mean_value=round(mean_value)
df_2["edad"].fillna(value=mean_value, inplace=True)
#Ahora si convertimos la columna edad a int
df_2["edad"]= df_2["edad"].astype(int)

#Llenamos los valores de estatura con el valor promedio
estaturas = pd.to_numeric(df_2.loc[:,'estatura'], errors='coerce')
estatura_promedio = round(estaturas.mean(), 2)
df_2['estatura'] = df_2['estatura'].replace(['No Especificado'], estatura_promedio)

#Eliminamos los valores "No Especificados" de la columna localidad, colocando en su lugar NAN y despues se elimina
df_2["localidad"]=df_2["localidad"].replace('No Especificado',np.NaN)
df_2=df_2.dropna(subset=["localidad"])




Eliminamos las columnas que no serán necesarias para responder las preguntas que nos realizamos, o bien que no aportan demasiada información:
*   La columna “país” tiene el mismo valor (México) para todos los registros de la lista
*   La columna etnia tiene 36133 registros con el valor “no especificado”
 de la lista

In [102]:
# Comprobamos si existen las columnas a eliminar

if set(['pais','etnia']).issubset(df_2.columns):
  df_2 = df_2.drop(['pais', 'etnia'], axis=1)


df_2.head(100)


fecha      hora           entidad              municipio  \
1   2007-02-05  02:00:26  Estado De Mexico         Valle De Bravo   
2   2009-03-14  08:30:54  Estado De Mexico                 Chalco   
3   2006-12-23  08:00:28  Estado De Mexico    Tlalnepantla De Baz   
4   2008-01-11  12:00:17  Estado De Mexico                 Toluca   
5   2012-05-27  12:00:43  Estado De Mexico         Nicolas Romero   
..         ...       ...               ...                    ...   
131 2012-02-09  12:00:00           Hidalgo        Pachuca De Soto   
132 2012-02-14  04:00:00           Hidalgo                Actopan   
133 2012-02-22  12:00:00           Hidalgo  Mineral De La Reforma   
134 2012-02-24  12:00:00           Hidalgo        Pachuca De Soto   
135 2012-02-22  12:00:00           Hidalgo  Mineral De La Reforma   

                 localidad nacionalidad estatura complexion    sexo  edad  \
1           Valle De Bravo     Mexicana      1.1    Delgada   Mujer     4   
2                   Chalco     Mexicana      1.1    Delgada  Hombre     6   
3      Tlalnepantla De Baz     Mexicana      1.1    Delgada  Hombre    10   
4                   Toluca     Mexicana      1.1    Delgada  Hombre     4   
5           Nicolas Romero     Mexicana      1.1    Delgada  Hombre     9   
..                     ...          ...      ...        ...     ...   ...   
131                Pachuca     Mexicana     1.55    Delgada   Mujer    13   
132                Actopan     Mexicana      1.7    Robusta   Mujer    16   
133  Mineral De La Reforma     Mexicana     1.56    Delgada   Mujer    15   
134                Pachuca     Mexicana     1.55    Mediana   Mujer    14   
135  Mineral De La Reforma     Mexicana      1.5    Mediana   Mujer    31   

                                    senas_particulares discapacidad  \
1                                      No Especificado      Ninguno   
2    Manchas, Parte Superior De Pierna Derecha, Cic...      Ninguno   
3    Lunares, A La Altura De La Cinturaforma De Man...      Ninguno   
4                                      No Especificado      Ninguno   
5                                      No Especificado      Ninguno   
..                                                 ...          ...   
131                                    No Especificado      Ninguno   
132                                    No Especificado      Ninguno   
133                                    No Especificado      Ninguno   
134                                    No Especificado      Ninguno   
135                                    Paño En La Cara      Ninguno   

          dependecia_origen  
1    Fgj - Estado De Mexico  
2    Fgj - Estado De Mexico  
3    Fgj - Estado De Mexico  
4    Fgj - Estado De Mexico  
5    Fgj - Estado De Mexico  
..                      ...  
131           Pgj - Hidalgo  
132           Pgj - Hidalgo  
133           Pgj - Hidalgo  
134           Pgj - Hidalgo  
135           Pgj - Hidalgo  

[100 rows x 13 columns]

En las columnas donde los valores “No Especificado” permiten el análisis que requerimos realizar se mantendran igual, esas columnas son:
*   Municipio
*   Nacionalidad
*   Complexión
*   Señas Particulares
*   Discapacidad


In [87]:
df_2.isna().sum()

fecha                 0
hora                  0
entidad               0
municipio             0
localidad             0
nacionalidad          0
estatura              0
complexion            0
sexo                  0
edad                  0
senas_particulares    0
discapacidad          0
dependecia_origen     0
dtype: int64

**Queremos saber ¿en que estado de la republica mexicana existen más sujetos extraviados?**

In [81]:
# CONSTRUIMOS NUEVOS DATA FRAMES PARA INTENTAR RESPONDER LAS PREGUNTAS PLANTEADAS
estados= pd.Series(df_2['entidad'], name='Estado')
cuenta = df_2.groupby(['entidad']).size()
df_freq_estados = pd.DataFrame(df_2.groupby(['entidad']).size())
df_freq_estados = df_freq_estados.rename(columns = {0:"frecuencia"})
# Ordenamos el Data Frame en forme descendente
df_freq_estados.sort_values('frecuencia', ascending=False, inplace=True)
print(f'El estado con más sujetos extraviados es {df_freq_estados.index[0]} con {df_freq_estados["frecuencia"][0]} ')

El estado con más sujetos extraviados es Tamaulipas con 5987 


Ahora presentamos completo el data frame con todas las entidades y numeros de sujetos extraviados

In [82]:
df_freq_estados

frecuencia
entidad                         
Tamaulipas                  5987
Estado De Mexico            3890
Jalisco                     3361
Sinaloa                     3023
Nuevo Leon                  2896
Chihuahua                   2181
Sonora                      2149
Puebla                      2066
Coahuila De Zaragoza        1714
Guerrero                    1438
Michoacan                   1207
Baja California              865
Ciudad De Mexico             727
Guanajuato                   615
Colima                       592
Veracruz                     520
Zacatecas                    510
Durango                      415
Queretaro                    283
Morelos                      239
Aguascalientes               222
Oaxaca                       190
Hidalgo                      145
Nayarit                      138
Chiapas                      108
Yucatan                       99
San Luis Potosi               96
Tabasco                       67
Quintana Roo                  61
Baja California Sur           39
Campeche                      34
No Especificado               26
Tlaxcala                      24

**Queremos saber si ¿en México se extravian solo mexicanos?**

In [83]:
df_nac= pd.crosstab(index=df_2['nacionalidad'],  columns='recuento', colnames=['nacionalidad'])
print(f'En Méxio estan perdidas personas de {len(df_nac.index)} nacionalidades, dstribuidos de la siguiente manera: \n')
df_nac
#f=df_nac['recuento'].max()
#df3=df_nac.merge(df_nac,left_on=f, right_on= how='left')

En Méxio estan perdidas personas de 26 nacionalidades, dstribuidos de la siguiente manera: 



nacionalidad     recuento
nacionalidad             
Alemana                 1
Argentina               1
Brasileña               1
Canadiense              3
China                   1
Colombiana              9
Costaricense            1
Cubana                  3
Dominicana              1
Eritrea                 1
Española                3
Estadounidense        123
Francesa                1
Guatemalteco           11
Haitiana                2
Holandesa               1
Hondureña              24
Italiana                3
Mexicana            33692
Nicaraguense            1
No Especificado      2028
Panameña                1
Paraguaya               1
Peruana                 2
Salvadoreña             8
Venezolana              4